{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h2>C - Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import icecream as ic\n",
    "import numpy as np\n",
    "import plotly\n",
    "import matplotlib as plt\n",
    "import xml.etree.ElementTree as ET\n",
    "import pandas as pd\n",
    "import psycopg2\n",
    "import psycopg2.extras\n",
    "import csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "C1: small-molecule roles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acetazolamide\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'has_part': [['Diamox'],\n",
       "  ['Sodium acetazolamide'],\n",
       "  ['sodium [(5-acetamido-1,3,4-thiadiazol-2-yl)sulfonyl]azanide'],\n",
       "  ['N-(5-Sulfamoyl-1,3,4-thiadiazol-2-yl)acetamide monosodium salt']],\n",
       " 'is_conjugate_base_of': [['[(5-acetamido-1,3,4-thiadiazol-2-yl)sulfonyl]azanide'],\n",
       "  ['acetazolamide']]}"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#chebi database\n",
    "\n",
    "hostname = 'localhost'\n",
    "database = 'Chebi'\n",
    "username = 'postgres'\n",
    "pwd = 'wlf1115'\n",
    "port_id = 5432\n",
    "conn = None\n",
    "\n",
    "query = 'Acetazolamide'\n",
    "with psycopg2.connect(\n",
    "                    host = hostname,\n",
    "                    dbname = database,\n",
    "                    user = username,\n",
    "                    password = pwd,\n",
    "                    port = port_id) as conn:\n",
    "    with conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as cur:\n",
    "\n",
    "        query_script = '''SELECT type, final_id\n",
    "                        FROM relation\n",
    "                        WHERE init_id = (\n",
    "                            SELECT compound_id\n",
    "                            FROM names\n",
    "                            WHERE name IN (%s)\n",
    "                            )'''\n",
    "        cur.execute(query_script, (query,))\n",
    "        \n",
    "        type_id = {}\n",
    "        type_compound = {}\n",
    "        for record in cur.fetchall():\n",
    "            type_id[record[0]] = []\n",
    "            type_compound[record[0]] = []\n",
    "            type_id[record[0]] = record[1]\n",
    "\n",
    "\n",
    "        for k,v in type_id.items():\n",
    "            # query1 = '''SELECT name\n",
    "            #             FROM names\n",
    "            #             WHERE compound_id IN ({}\n",
    "            #             )'''.format(tuple([x]))\n",
    "            query_script1 = '''SELECT name\n",
    "                        FROM names\n",
    "                        WHERE compound_id IN (%s)\n",
    "                        '''\n",
    "            query1 = tuple([v])\n",
    "            cur.execute(query_script1, query1)\n",
    "            for record in cur.fetchall():\n",
    "                type_compound[k].append(record)\n",
    "\n",
    "print(query)\n",
    "type_compound    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "C2: metabolic pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://www.ebi.ac.uk/biomodels/MODEL1109130000#Files #!how to get binary data, how is the graph represented # read paper\n",
    "# We downloaded the reconstruction of human metabolism (Recon)57 from Pathway Commons58 (http://www.pathwaycommons.org, July 2017) in binary interaction form. \n",
    "# Data were represented as an undirected graph where nodes are metabolites and edges denote reactions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>init</th>\n",
       "      <th>process</th>\n",
       "      <th>affected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>B3GALNT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>B3GNT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>controls-production-of</td>\n",
       "      <td>CHEBI:58223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>ENTPD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>NME1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52598</th>\n",
       "      <td>ZADH2</td>\n",
       "      <td>controls-production-of</td>\n",
       "      <td>CHEBI:17790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52599</th>\n",
       "      <td>ZADH2</td>\n",
       "      <td>controls-production-of</td>\n",
       "      <td>CHEBI:18041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52600</th>\n",
       "      <td>ZADH2</td>\n",
       "      <td>controls-production-of</td>\n",
       "      <td>CHEBI:28972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52601</th>\n",
       "      <td>ZADH2</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>GLO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52602</th>\n",
       "      <td>ZADH2</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>HPRT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>52603 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         init                 process     affected\n",
       "0      A4GALT      catalysis-precedes     B3GALNT1\n",
       "1      A4GALT      catalysis-precedes       B3GNT3\n",
       "2      A4GALT  controls-production-of  CHEBI:58223\n",
       "3      A4GALT      catalysis-precedes       ENTPD4\n",
       "4      A4GALT      catalysis-precedes         NME1\n",
       "...       ...                     ...          ...\n",
       "52598   ZADH2  controls-production-of  CHEBI:17790\n",
       "52599   ZADH2  controls-production-of  CHEBI:18041\n",
       "52600   ZADH2  controls-production-of  CHEBI:28972\n",
       "52601   ZADH2      catalysis-precedes         GLO1\n",
       "52602   ZADH2      catalysis-precedes        HPRT1\n",
       "\n",
       "[52603 rows x 3 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recon_df = pd.read_csv(\"../PathwayCommons.8.reconx.BINARY_SIF.hgnc.txt\", sep='\\t', names=['init', 'process','affected'])\n",
    "recon_df #how to convert protein name to abbreviation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "C3: signaling pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#  C3 space (and the C4 and C5 spaces) is aimed at any molecule with known protein targets. \n",
    "# In this case, we list the biological pathways that may be affected by the interaction of a molecule with its targets. \n",
    "# Human pathways were collected from Reactome (https://reactome.org, May 2017) #! use of graphDB\n",
    "\n",
    "# we chose to use binding activities from B4, since this is an extensive dataset containing mostly literature data with well-accepted activity thresholds. \n",
    "# In B4, 24.5% of the compound–protein interactions do not correspond to human proteins. #!need B4 data\n",
    "# These were mapped to their human orthologs using MetaPhOrs60 (http://orthology.phylomedb.org, May 2017), following the observation that binding activities can be safely transferred between orthologous proteins61,  #!how is it done, it is a database?\n",
    "# especially if they belong to closely related species, as it is the case for B4 data62. Of all the nonhuman proteins mapped to the human orthologs, 94.4% were mammal proteins #!http://orthology.phylomedb.org/download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/78/zn54_c9d2jb_8j8xx5s59dqw0000gn/T/ipykernel_1245/1950680321.py:4: DtypeWarning: Columns (5,7,8,9,10,11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  chemble_B1 = pd.read_csv(\"../chembl31forlinfeng/B1_data.csv\", sep=',', header=0)\n"
     ]
    }
   ],
   "source": [
    "reactome_df = pd.read_csv(\"../UniProt2Reactome_All_Levels.txt\", sep='\\t', names=['uniprot_id','interaction_code','web','pathway','IEA','organism'])\n",
    "reactome_df = reactome_df[['uniprot_id','pathway']]\n",
    "\n",
    "chemble_B1 = pd.read_csv(\"../chembl31forlinfeng/B1_data.csv\", sep=',', header=0)\n",
    "chemble_B1_target_uniprot = chemble_B1[chemble_B1['compound_name']=='Acetazolamide']['uniprot_id']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>412824</th>\n",
       "      <td>O00255</td>\n",
       "      <td>Signal Transduction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412825</th>\n",
       "      <td>O00255</td>\n",
       "      <td>Signaling by TGF-beta Receptor Complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412826</th>\n",
       "      <td>O00255</td>\n",
       "      <td>Signaling by Rho GTPases</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412827</th>\n",
       "      <td>O00255</td>\n",
       "      <td>RHO GTPase Effectors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412828</th>\n",
       "      <td>O00255</td>\n",
       "      <td>Signaling by WNT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903808</th>\n",
       "      <td>Q9Y6L6</td>\n",
       "      <td>Disorders of transmembrane transporters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903809</th>\n",
       "      <td>Q9Y6L6</td>\n",
       "      <td>Transport of organic anions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903810</th>\n",
       "      <td>Q9Y6L6</td>\n",
       "      <td>Metabolism of steroids</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903811</th>\n",
       "      <td>Q9Y6L6</td>\n",
       "      <td>Drug ADME</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903812</th>\n",
       "      <td>Q9Y6L6</td>\n",
       "      <td>Atorvastatin ADME</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2439 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       uniprot_id                                  pathway\n",
       "412824     O00255                      Signal Transduction\n",
       "412825     O00255   Signaling by TGF-beta Receptor Complex\n",
       "412826     O00255                 Signaling by Rho GTPases\n",
       "412827     O00255                     RHO GTPase Effectors\n",
       "412828     O00255                         Signaling by WNT\n",
       "...           ...                                      ...\n",
       "903808     Q9Y6L6  Disorders of transmembrane transporters\n",
       "903809     Q9Y6L6              Transport of organic anions\n",
       "903810     Q9Y6L6                   Metabolism of steroids\n",
       "903811     Q9Y6L6                                Drug ADME\n",
       "903812     Q9Y6L6                        Atorvastatin ADME\n",
       "\n",
       "[2439 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#  biological pathways that may be affected by the interaction of a molecule with its targets. \n",
    "# ! still need to make sure all the non human genes are turned into human orthologs\n",
    "reactome_df[reactome_df['uniprot_id'].isin(chemble_B1_target_uniprot)]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "C4: biological processes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'DB': 'UniProtKB', 'DB_Object_ID': 'A0A024QZ33', 'DB_Object_Symbol': 'NSRP1', 'Qualifier': ['involved_in'], 'GO_ID': 'GO:0000381', 'DB:Reference': ['GO_REF:0000002'], 'Evidence': 'IEA', 'With': ['InterPro:IPR018612'], 'Aspect': 'P', 'DB_Object_Name': 'Nuclear speckle splicing regulatory protein 1', 'Synonym': ['NSRP1', 'CCDC55', 'hCG_1646942'], 'DB_Object_Type': 'protein', 'Taxon_ID': ['9606'], 'Date': '20220907', 'Assigned_By': 'InterPro', 'Annotation_Extension': '', 'Gene_Product_Form_ID': ''}\n"
     ]
    }
   ],
   "source": [
    "with open('../GOA/25.H_sapiens.goa', 'r') as handle:\n",
    "\n",
    "    for rec in gafiterator(handle):\n",
    "        print(rec)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0004064\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0004089\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0004089\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0004089\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0004089\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0004089\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0004089\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0005515\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0005515\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0008270\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0016829\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0016836\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0016836\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0018820\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['enables']\n",
      "GO:0046872\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['involved_in']\n",
      "GO:0006730\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['involved_in']\n",
      "GO:0006730\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['is_active_in']\n",
      "GO:0005737\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0005737\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0005737\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0005737\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0005829\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0005829\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0005829\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0005829\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0005829\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0005829\n",
      "====================\n",
      "P00915\n",
      "Carbonic anhydrase 1\n",
      "['located_in']\n",
      "GO:0070062\n",
      "====================\n"
     ]
    }
   ],
   "source": [
    "# We downloaded the Gene Ontology Annotation database (https://www.ebi.ac.uk/GOA, May 2017)\n",
    "# read the ‘biological process’ branch of the ontology as a directed acyclic graph (‘is a’ relationships). \n",
    "# Proteins were annotated with their Gene Ontology Annotation biological process terms plus parent terms (up to the root of the directed acyclic graph). \n",
    "# Similar to C3, we associated molecules with biological process terms by simply checking the annotations of the molecule targets (B4).\n",
    "from ast import And\n",
    "from Bio.UniProt.GOA import gafiterator, record_has\n",
    "\n",
    "for x in chemble_B1_target_uniprot.to_list():\n",
    "    Interpro_ID = {'With': set([f'InterPro:IPR018612'])} \n",
    "    DB_Object_ID = {'DB_Object_ID': set([x])} \n",
    "\n",
    "    # DB_Object_name = {'DB_Object_name' : set(['Nuclear speckle splicing regulatory protein 1'])} \n",
    "\n",
    "    with open('../GOA/25.H_sapiens.goa', 'r') as handle:\n",
    "\n",
    "        for rec in gafiterator(handle):\n",
    "            if record_has(rec, DB_Object_ID):\n",
    "                for key in ('DB_Object_ID','DB_Object_Name', 'Qualifier', 'GO_ID'):\n",
    "                    print(rec[key])\n",
    "                print('='*20)\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "C5: interactomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We collected five representative PPI networks, \n",
    "# namely STRING (score of >700, that is, high confidence)63 (v.10, https://string-db.org) (14,725 proteins (p), 300,686 interactions (i)), #! waiting for data\n",
    "# InWeb (score of 0.5)64 (http://www.intomics.com/inbio/map, March 2017) (10,100 p, 168,970 i), #!paid service? <- not gonna use\n",
    "# a portion of Pathway Commons containing interactions from known pathways (Kyoto Encylopedia of Genes and Genomes65, NetPath66, PANTHER67 and WikiPathways68) (9,344 p, 242,962 i), #!need to find out\n",
    "# an in-house network of physical binary PPIs69 (13,038 p, 64,659 i),  #!need to find in house pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "stringDB_uniprot = pd.read_csv('../human_HighConfidence_string_PPI_add_uniprot.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>protein1_uniprot_id</th>\n",
       "      <th>protein1_protein_name</th>\n",
       "      <th>protein2_uniprot_id</th>\n",
       "      <th>protein2_protein_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000324287</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>736</td>\n",
       "      <td>767</td>\n",
       "      <td>P84085</td>\n",
       "      <td>ADP-ribosylation factor 5</td>\n",
       "      <td>Q15057</td>\n",
       "      <td>Arf-GAP with coiled-coil, ANK repeat and PH do...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000387286</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>248</td>\n",
       "      <td>600</td>\n",
       "      <td>141</td>\n",
       "      <td>730</td>\n",
       "      <td>P84085</td>\n",
       "      <td>ADP-ribosylation factor 5</td>\n",
       "      <td>P62820</td>\n",
       "      <td>Ras-related protein Rab-1A (EC 3.6.5.2) (YPT1-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000262812</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>190</td>\n",
       "      <td>163</td>\n",
       "      <td>600</td>\n",
       "      <td>173</td>\n",
       "      <td>745</td>\n",
       "      <td>P84085</td>\n",
       "      <td>ADP-ribosylation factor 5</td>\n",
       "      <td>O14579</td>\n",
       "      <td>Coatomer subunit epsilon (Epsilon-coat protein...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000158762</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>736</td>\n",
       "      <td>765</td>\n",
       "      <td>P84085</td>\n",
       "      <td>ADP-ribosylation factor 5</td>\n",
       "      <td>Q15027</td>\n",
       "      <td>Arf-GAP with coiled-coil, ANK repeat and PH do...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000449270</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>162</td>\n",
       "      <td>210</td>\n",
       "      <td>600</td>\n",
       "      <td>194</td>\n",
       "      <td>757</td>\n",
       "      <td>P84085</td>\n",
       "      <td>ADP-ribosylation factor 5</td>\n",
       "      <td>P61923</td>\n",
       "      <td>Coatomer subunit zeta-1 (Zeta-1-coat protein) ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               protein1              protein2  neighborhood  fusion  \\\n",
       "0  9606.ENSP00000000233  9606.ENSP00000324287             0       0   \n",
       "1  9606.ENSP00000000233  9606.ENSP00000387286             0       0   \n",
       "2  9606.ENSP00000000233  9606.ENSP00000262812             0       0   \n",
       "3  9606.ENSP00000000233  9606.ENSP00000158762             0       0   \n",
       "4  9606.ENSP00000000233  9606.ENSP00000449270             0       0   \n",
       "\n",
       "   cooccurence  coexpression  experimental  database  textmining  \\\n",
       "0            0            49           147         0         736   \n",
       "1            0            79           248       600         141   \n",
       "2            0           190           163       600         173   \n",
       "3            0             0           147         0         736   \n",
       "4            0           162           210       600         194   \n",
       "\n",
       "   combined_score protein1_uniprot_id      protein1_protein_name  \\\n",
       "0             767              P84085  ADP-ribosylation factor 5   \n",
       "1             730              P84085  ADP-ribosylation factor 5   \n",
       "2             745              P84085  ADP-ribosylation factor 5   \n",
       "3             765              P84085  ADP-ribosylation factor 5   \n",
       "4             757              P84085  ADP-ribosylation factor 5   \n",
       "\n",
       "  protein2_uniprot_id                              protein2_protein_name  \n",
       "0              Q15057  Arf-GAP with coiled-coil, ANK repeat and PH do...  \n",
       "1              P62820  Ras-related protein Rab-1A (EC 3.6.5.2) (YPT1-...  \n",
       "2              O14579  Coatomer subunit epsilon (Epsilon-coat protein...  \n",
       "3              Q15027  Arf-GAP with coiled-coil, ANK repeat and PH do...  \n",
       "4              P61923  Coatomer subunit zeta-1 (Zeta-1-coat protein) ...  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stringDB_uniprot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "stringDB_df = pd.read_csv('../human_HighConfidence_string_PPI.csv', sep=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create input for hotnet -stringDB\n",
    "bla = np.append(stringDB_df['protein1'].unique(), stringDB_df['protein2'].unique()) #this needs to be done as there are unique prot in prot2 that don't exist in prot1\n",
    "bla = set(bla.tolist())\n",
    "\n",
    "protein_encoding = {}\n",
    "\n",
    "for i, x in enumerate(bla):\n",
    "    protein_encoding[x] = i+1\n",
    "protein1_encoding = []\n",
    "for x in stringDB_df['protein1']:\n",
    "    protein1_encoding.append(protein_encoding[x])\n",
    "    \n",
    "protein2_encoding = []\n",
    "for x in stringDB_df['protein2']:\n",
    "    protein2_encoding.append(protein_encoding[x])\n",
    "    \n",
    "stringDB_df['protein1_encoding'] = protein1_encoding\n",
    "stringDB_df['protein2_encoding'] = protein2_encoding\n",
    "\n",
    "protein_encoding = pd.DataFrame({'Names':list(protein_encoding.keys()), 'Encodings':list(protein_encoding.values())})\n",
    "\n",
    "#export index file\n",
    "protein_encoding[['Encodings','Names']].to_csv('stringDB_index.txt', sep = \" \", index = False, index_label = False, header = False)\n",
    "stringDB_df[['protein1_encoding','protein2_encoding']].to_csv('stringDB_edge.txt', sep = \" \", index = False, index_label = False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein1</th>\n",
       "      <th>protein2</th>\n",
       "      <th>neighborhood</th>\n",
       "      <th>fusion</th>\n",
       "      <th>cooccurence</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimental</th>\n",
       "      <th>database</th>\n",
       "      <th>textmining</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>protein1_encoding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000324287</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>736</td>\n",
       "      <td>767</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000387286</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>248</td>\n",
       "      <td>600</td>\n",
       "      <td>141</td>\n",
       "      <td>730</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000262812</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>190</td>\n",
       "      <td>163</td>\n",
       "      <td>600</td>\n",
       "      <td>173</td>\n",
       "      <td>745</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000158762</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>147</td>\n",
       "      <td>0</td>\n",
       "      <td>736</td>\n",
       "      <td>765</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000000233</td>\n",
       "      <td>9606.ENSP00000449270</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>162</td>\n",
       "      <td>210</td>\n",
       "      <td>600</td>\n",
       "      <td>194</td>\n",
       "      <td>757</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504021</th>\n",
       "      <td>9606.ENSP00000485663</td>\n",
       "      <td>9606.ENSP00000248342</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>699</td>\n",
       "      <td>996</td>\n",
       "      <td>900</td>\n",
       "      <td>957</td>\n",
       "      <td>999</td>\n",
       "      <td>504022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504022</th>\n",
       "      <td>9606.ENSP00000485663</td>\n",
       "      <td>9606.ENSP00000416255</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>808</td>\n",
       "      <td>0</td>\n",
       "      <td>473</td>\n",
       "      <td>908</td>\n",
       "      <td>504023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504023</th>\n",
       "      <td>9606.ENSP00000485663</td>\n",
       "      <td>9606.ENSP00000220849</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>907</td>\n",
       "      <td>993</td>\n",
       "      <td>900</td>\n",
       "      <td>896</td>\n",
       "      <td>999</td>\n",
       "      <td>504024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504024</th>\n",
       "      <td>9606.ENSP00000485663</td>\n",
       "      <td>9606.ENSP00000346088</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>711</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>738</td>\n",
       "      <td>504025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504025</th>\n",
       "      <td>9606.ENSP00000485663</td>\n",
       "      <td>9606.ENSP00000341885</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>525</td>\n",
       "      <td>821</td>\n",
       "      <td>600</td>\n",
       "      <td>58</td>\n",
       "      <td>963</td>\n",
       "      <td>504026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>504026 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    protein1              protein2  neighborhood  fusion  \\\n",
       "0       9606.ENSP00000000233  9606.ENSP00000324287             0       0   \n",
       "1       9606.ENSP00000000233  9606.ENSP00000387286             0       0   \n",
       "2       9606.ENSP00000000233  9606.ENSP00000262812             0       0   \n",
       "3       9606.ENSP00000000233  9606.ENSP00000158762             0       0   \n",
       "4       9606.ENSP00000000233  9606.ENSP00000449270             0       0   \n",
       "...                      ...                   ...           ...     ...   \n",
       "504021  9606.ENSP00000485663  9606.ENSP00000248342             0       0   \n",
       "504022  9606.ENSP00000485663  9606.ENSP00000416255             0       0   \n",
       "504023  9606.ENSP00000485663  9606.ENSP00000220849             0       0   \n",
       "504024  9606.ENSP00000485663  9606.ENSP00000346088             0       0   \n",
       "504025  9606.ENSP00000485663  9606.ENSP00000341885             0       0   \n",
       "\n",
       "        cooccurence  coexpression  experimental  database  textmining  \\\n",
       "0                 0            49           147         0         736   \n",
       "1                 0            79           248       600         141   \n",
       "2                 0           190           163       600         173   \n",
       "3                 0             0           147         0         736   \n",
       "4                 0           162           210       600         194   \n",
       "...             ...           ...           ...       ...         ...   \n",
       "504021            0           699           996       900         957   \n",
       "504022            0           167           808         0         473   \n",
       "504023            0           907           993       900         896   \n",
       "504024            0           711             0         0         131   \n",
       "504025            0           525           821       600          58   \n",
       "\n",
       "        combined_score  protein1_encoding  \n",
       "0                  767                  1  \n",
       "1                  730                  2  \n",
       "2                  745                  3  \n",
       "3                  765                  4  \n",
       "4                  757                  5  \n",
       "...                ...                ...  \n",
       "504021             999             504022  \n",
       "504022             908             504023  \n",
       "504023             999             504024  \n",
       "504024             738             504025  \n",
       "504025             963             504026  \n",
       "\n",
       "[504026 rows x 11 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stringDB_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9606.ENSP00000324287\n",
      "207062    207063\n",
      "207063    207064\n",
      "207064    207065\n",
      "207065    207066\n",
      "207066    207067\n",
      "Name: protein1_encoding, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "protein2_encoding = []\n",
    "for x in stringDB_df['protein2']:\n",
    "    print(x)\n",
    "    print(stringDB_df[stringDB_df['protein1']==x]['protein1_encoding'])\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "stringDB_df = stringDB_df[stringDB_df['combined_score']>700]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Intersection(lst1, lst2):\n",
    "    return set(lst1).intersection(lst2)\n",
    "     \n",
    "# Driver Code\n",
    "\n",
    "print(Intersection(stringDB_df['protein1'].unique(), stringDB_df['protein2'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# scaling to between 0 and 10 - https://stats.stackexchange.com/questions/281162/scale-a-number-between-a-range\n",
    "r_min = 90 #minimum of the range of your measurement\n",
    "r_max = 900 #maximum of the range of your measurement\n",
    "input_max = 10 #maximum of the range of your desired target scaling\n",
    "input_min = 0 #minimum of the range of your desired target scaling\n",
    "input_value = 95\n",
    "int((input_value - r_min) / (r_max - r_min) * (input_max - input_min) + input_min)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['protein1 protein2 neighborhood fusion cooccurence coexpression experimental database textmining combined_score'], dtype='object')"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PROT1</th>\n",
       "      <th>PROT2</th>\n",
       "      <th>RANK_MAJOR</th>\n",
       "      <th>RANK_MINOR</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>PDB_ID</th>\n",
       "      <th>BIO_UNIT</th>\n",
       "      <th>CHAIN1</th>\n",
       "      <th>MODEL1</th>\n",
       "      <th>SEQ_IDENT1</th>\n",
       "      <th>...</th>\n",
       "      <th>SEQ_END1</th>\n",
       "      <th>DOMAIN1</th>\n",
       "      <th>CHAIN2</th>\n",
       "      <th>MODEL2</th>\n",
       "      <th>SEQ_IDENT2</th>\n",
       "      <th>COVERAGE2</th>\n",
       "      <th>SEQ_BEGIN2</th>\n",
       "      <th>SEQ_END2</th>\n",
       "      <th>DOMAIN2</th>\n",
       "      <th>FILENAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A024RAV5</td>\n",
       "      <td>P02647</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>2mse</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>185</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>74.2</td>\n",
       "      <td>68</td>\n",
       "      <td>265</td>\n",
       "      <td>-</td>\n",
       "      <td>A0A024RAV5-P02647-EXP-2mse.pdb1-B-0-A-0.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A024RAV5</td>\n",
       "      <td>P02647</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>6pts</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>185</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>74.2</td>\n",
       "      <td>68</td>\n",
       "      <td>265</td>\n",
       "      <td>-</td>\n",
       "      <td>A0A024RAV5-P02647-EXP-6pts.pdb1-B-0-A-0.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A024RAV5</td>\n",
       "      <td>P10398</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>2mse</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>185</td>\n",
       "      <td>-</td>\n",
       "      <td>D</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>19</td>\n",
       "      <td>91</td>\n",
       "      <td>-</td>\n",
       "      <td>A0A024RAV5-P10398-EXP-2mse.pdb1-B-0-D-0.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A075B5G3</td>\n",
       "      <td>Q8NBP7</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>4ov6</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>99</td>\n",
       "      <td>-</td>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>95.6</td>\n",
       "      <td>42.5</td>\n",
       "      <td>153</td>\n",
       "      <td>446</td>\n",
       "      <td>-</td>\n",
       "      <td>A0A075B5G3-Q8NBP7-EXP-4ov6.pdb1-F-0-B-0.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A0A075B5G3</td>\n",
       "      <td>Q8NBP7</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>4ov6</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>96</td>\n",
       "      <td>-</td>\n",
       "      <td>E</td>\n",
       "      <td>0</td>\n",
       "      <td>92.2</td>\n",
       "      <td>42.3</td>\n",
       "      <td>153</td>\n",
       "      <td>445</td>\n",
       "      <td>-</td>\n",
       "      <td>A0A075B5G3-Q8NBP7-EXP-4ov6.pdb2-G-0-E-0.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124919</th>\n",
       "      <td>V9HW83</td>\n",
       "      <td>V9HW83</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>4wb9</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.4</td>\n",
       "      <td>9</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>V9HW83-V9HW83-EXP-4wb9.pdb1-A-2-A-3.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124920</th>\n",
       "      <td>V9HW83</td>\n",
       "      <td>V9HW83</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>4wb9</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.4</td>\n",
       "      <td>9</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>V9HW83-V9HW83-EXP-4wb9.pdb1-A-0-A-1.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124921</th>\n",
       "      <td>V9HW83</td>\n",
       "      <td>V9HW83</td>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>4wb9</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.4</td>\n",
       "      <td>9</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>V9HW83-V9HW83-EXP-4wb9.pdb1-A-0-A-2.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124922</th>\n",
       "      <td>V9HW83</td>\n",
       "      <td>V9HW83</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>4wb9</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.4</td>\n",
       "      <td>9</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>V9HW83-V9HW83-EXP-4wb9.pdb1-A-0-A-3.pdb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124923</th>\n",
       "      <td>V9HW83</td>\n",
       "      <td>V9HW83</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>Structure</td>\n",
       "      <td>4wb9</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>98.4</td>\n",
       "      <td>9</td>\n",
       "      <td>501</td>\n",
       "      <td>-</td>\n",
       "      <td>V9HW83-V9HW83-EXP-4wb9.pdb1-A-1-A-3.pdb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>124923 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             PROT1   PROT2 RANK_MAJOR RANK_MINOR       TYPE PDB_ID BIO_UNIT  \\\n",
       "1       A0A024RAV5  P02647          1          0  Structure   2mse        1   \n",
       "2       A0A024RAV5  P02647          2          0  Structure   6pts        1   \n",
       "3       A0A024RAV5  P10398          1          0  Structure   2mse        1   \n",
       "4       A0A075B5G3  Q8NBP7          1          0  Structure   4ov6        1   \n",
       "5       A0A075B5G3  Q8NBP7          2          0  Structure   4ov6        2   \n",
       "...            ...     ...        ...        ...        ...    ...      ...   \n",
       "124919      V9HW83  V9HW83         54          0  Structure   4wb9        1   \n",
       "124920      V9HW83  V9HW83         55          0  Structure   4wb9        1   \n",
       "124921      V9HW83  V9HW83         56          0  Structure   4wb9        1   \n",
       "124922      V9HW83  V9HW83         57          0  Structure   4wb9        1   \n",
       "124923      V9HW83  V9HW83         58          0  Structure   4wb9        1   \n",
       "\n",
       "       CHAIN1 MODEL1 SEQ_IDENT1  ... SEQ_END1 DOMAIN1 CHAIN2 MODEL2  \\\n",
       "1           B      0      100.0  ...      185       -      A      0   \n",
       "2           B      0      100.0  ...      185       -      A      0   \n",
       "3           B      0      100.0  ...      185       -      D      0   \n",
       "4           F      0      100.0  ...       99       -      B      0   \n",
       "5           G      0      100.0  ...       96       -      E      0   \n",
       "...       ...    ...        ...  ...      ...     ...    ...    ...   \n",
       "124919      A      2      100.0  ...      501       -      A      3   \n",
       "124920      A      0      100.0  ...      501       -      A      1   \n",
       "124921      A      0      100.0  ...      501       -      A      2   \n",
       "124922      A      0      100.0  ...      501       -      A      3   \n",
       "124923      A      1      100.0  ...      501       -      A      3   \n",
       "\n",
       "       SEQ_IDENT2 COVERAGE2 SEQ_BEGIN2 SEQ_END2 DOMAIN2  \\\n",
       "1           100.0      74.2         68      265       -   \n",
       "2           100.0      74.2         68      265       -   \n",
       "3           100.0      12.0         19       91       -   \n",
       "4            95.6      42.5        153      446       -   \n",
       "5            92.2      42.3        153      445       -   \n",
       "...           ...       ...        ...      ...     ...   \n",
       "124919      100.0      98.4          9      501       -   \n",
       "124920      100.0      98.4          9      501       -   \n",
       "124921      100.0      98.4          9      501       -   \n",
       "124922      100.0      98.4          9      501       -   \n",
       "124923      100.0      98.4          9      501       -   \n",
       "\n",
       "                                           FILENAME  \n",
       "1       A0A024RAV5-P02647-EXP-2mse.pdb1-B-0-A-0.pdb  \n",
       "2       A0A024RAV5-P02647-EXP-6pts.pdb1-B-0-A-0.pdb  \n",
       "3       A0A024RAV5-P10398-EXP-2mse.pdb1-B-0-D-0.pdb  \n",
       "4       A0A075B5G3-Q8NBP7-EXP-4ov6.pdb1-F-0-B-0.pdb  \n",
       "5       A0A075B5G3-Q8NBP7-EXP-4ov6.pdb2-G-0-E-0.pdb  \n",
       "...                                             ...  \n",
       "124919      V9HW83-V9HW83-EXP-4wb9.pdb1-A-2-A-3.pdb  \n",
       "124920      V9HW83-V9HW83-EXP-4wb9.pdb1-A-0-A-1.pdb  \n",
       "124921      V9HW83-V9HW83-EXP-4wb9.pdb1-A-0-A-2.pdb  \n",
       "124922      V9HW83-V9HW83-EXP-4wb9.pdb1-A-0-A-3.pdb  \n",
       "124923      V9HW83-V9HW83-EXP-4wb9.pdb1-A-1-A-3.pdb  \n",
       "\n",
       "[124923 rows x 22 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# an in-house network of physical binary PPIs69 (13,038 p, 64,659 i),  #!need to find in house pipeline\n",
    "import csv\n",
    "# read flash.dat to a list of lists\n",
    "datContent = [i.strip().split() for i in open(\"../interactions_interactome_3d.dat\").readlines()]\n",
    "interaction3D_df = pd.DataFrame(datContent, columns = datContent[0])\n",
    "interaction3D_df = interaction3D_df[1:] \n",
    "interaction3D_df #! how did you get binary PPI interactome info from this?  https://interactome3d.irbbarcelona.org/help.php#interaction_page\n",
    "# https://interactome3d.irbbarcelona.org/help.php#interaction_page"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "bla = np.append(interaction3D_df['PROT1'].unique(), interaction3D_df['PROT2'].unique()) #this needs to be done as there are unique prot in prot2 that don't exist in prot1\n",
    "bla = set(bla.tolist())\n",
    "\n",
    "protein_encoding = {}\n",
    "for i, x in enumerate(bla):\n",
    "    protein_encoding[x] = i+1\n",
    "protein1_encoding = []\n",
    "for x in interaction3D_df['PROT1']:\n",
    "    protein1_encoding.append(protein_encoding[x])\n",
    "    \n",
    "protein2_encoding = []\n",
    "for x in interaction3D_df['PROT2']:\n",
    "    protein2_encoding.append(protein_encoding[x])\n",
    "    \n",
    "interaction3D_df['PROT1_encoding'] = protein1_encoding\n",
    "interaction3D_df['PROT2_encoding'] = protein2_encoding\n",
    "\n",
    "protein_encoding = pd.DataFrame({'Names':list(protein_encoding.keys()), 'Encodings':list(protein_encoding.values())})\n",
    "\n",
    "#export index file\n",
    "protein_encoding[['Encodings','Names']].to_csv('interaction3D_index.txt', sep = \" \", index = False, index_label = False, header = False)\n",
    "interaction3D_df[['PROT1_encoding','PROT2_encoding']].to_csv('interaction3D_edge.txt', sep = \" \", index = False, index_label = False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'read' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [103], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mgseapy\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m read\u001b[39m.\u001b[39mgmt(\u001b[39m'\u001b[39m\u001b[39m../m2.cp.wikipathways.v2022.1.Mm.symbols.gmt\u001b[39m\u001b[39m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'read' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "# a portion of Pathway Commons containing interactions from known pathways (Kyoto Encylopedia of Genes and Genomes65, NetPath66, PANTHER67 and WikiPathways68) (9,344 p, 242,962 i), #!need to find out\n",
    "\n",
    "import gseapy\n",
    "read.gmt('../m2.cp.wikipathways.v2022.1.Mm.symbols.gmt')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../m2.cp.wikipathways.v2022.1.Mm.symbols.gmt') as gmt:\n",
    "    l = (gmt.read().split('\\t'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<gseapy.enrichr.Enrichr at 0x170fce2c0>"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gseapy.enrichr(gene_list=l,gene_sets='KEGG_2016', outdir='test')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PARTICIPANT_A</th>\n",
       "      <th>INTERACTION_TYPE</th>\n",
       "      <th>PARTICIPANT_B</th>\n",
       "      <th>INTERACTION_DATA_SOURCE</th>\n",
       "      <th>INTERACTION_PUBMED_ID</th>\n",
       "      <th>PATHWAY_NAMES</th>\n",
       "      <th>MEDIATOR_IDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABAT</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>ALDH5A1</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gamma-aminobutyric acid synthesis</td>\n",
       "      <td>http://pathwaycommons.org/pc13/Catalysis_c5d26...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABAT</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>ALDH6A1</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pyrimidine Metabolism</td>\n",
       "      <td>http://pathwaycommons.org/pc13/Catalysis_80d0d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABAT</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>CSAD</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gamma-aminobutyric acid synthesis</td>\n",
       "      <td>http://pathwaycommons.org/pc13/Catalysis_c5d26...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABAT</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>GAD1</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gamma-aminobutyric acid synthesis</td>\n",
       "      <td>http://pathwaycommons.org/pc13/Catalysis_c5d26...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABAT</td>\n",
       "      <td>catalysis-precedes</td>\n",
       "      <td>GAD2</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gamma-aminobutyric acid synthesis</td>\n",
       "      <td>http://pathwaycommons.org/pc13/Catalysis_c5d26...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32985</th>\n",
       "      <td>BAX</td>\n",
       "      <td>ProteinReference</td>\n",
       "      <td>BAX_HUMAN</td>\n",
       "      <td>uniprot knowledgebase:Q07812</td>\n",
       "      <td>hgnc symbol:BAX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32986</th>\n",
       "      <td>SPRY2</td>\n",
       "      <td>ProteinReference</td>\n",
       "      <td>SPY2_HUMAN</td>\n",
       "      <td>uniprot knowledgebase:O43597</td>\n",
       "      <td>hgnc symbol:SPRY2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32987</th>\n",
       "      <td>SPRY1</td>\n",
       "      <td>ProteinReference</td>\n",
       "      <td>SPY1_HUMAN</td>\n",
       "      <td>uniprot knowledgebase:O43609</td>\n",
       "      <td>hgnc symbol:SPRY1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32988</th>\n",
       "      <td>RGS12</td>\n",
       "      <td>ProteinReference</td>\n",
       "      <td>RGS12_HUMAN</td>\n",
       "      <td>uniprot knowledgebase:O14924</td>\n",
       "      <td>hgnc symbol:RGS12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32989</th>\n",
       "      <td>RGS11</td>\n",
       "      <td>ProteinReference</td>\n",
       "      <td>RGS11_HUMAN</td>\n",
       "      <td>uniprot knowledgebase:O94810</td>\n",
       "      <td>hgnc symbol:RGS11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32990 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      PARTICIPANT_A    INTERACTION_TYPE PARTICIPANT_B  \\\n",
       "0              ABAT  catalysis-precedes       ALDH5A1   \n",
       "1              ABAT  catalysis-precedes       ALDH6A1   \n",
       "2              ABAT  catalysis-precedes          CSAD   \n",
       "3              ABAT  catalysis-precedes          GAD1   \n",
       "4              ABAT  catalysis-precedes          GAD2   \n",
       "...             ...                 ...           ...   \n",
       "32985           BAX    ProteinReference     BAX_HUMAN   \n",
       "32986         SPRY2    ProteinReference    SPY2_HUMAN   \n",
       "32987         SPRY1    ProteinReference    SPY1_HUMAN   \n",
       "32988         RGS12    ProteinReference   RGS12_HUMAN   \n",
       "32989         RGS11    ProteinReference   RGS11_HUMAN   \n",
       "\n",
       "            INTERACTION_DATA_SOURCE INTERACTION_PUBMED_ID  \\\n",
       "0                           PANTHER                   NaN   \n",
       "1                           PANTHER                   NaN   \n",
       "2                           PANTHER                   NaN   \n",
       "3                           PANTHER                   NaN   \n",
       "4                           PANTHER                   NaN   \n",
       "...                             ...                   ...   \n",
       "32985  uniprot knowledgebase:Q07812       hgnc symbol:BAX   \n",
       "32986  uniprot knowledgebase:O43597     hgnc symbol:SPRY2   \n",
       "32987  uniprot knowledgebase:O43609     hgnc symbol:SPRY1   \n",
       "32988  uniprot knowledgebase:O14924     hgnc symbol:RGS12   \n",
       "32989  uniprot knowledgebase:O94810     hgnc symbol:RGS11   \n",
       "\n",
       "                           PATHWAY_NAMES  \\\n",
       "0      Gamma-aminobutyric acid synthesis   \n",
       "1                  Pyrimidine Metabolism   \n",
       "2      Gamma-aminobutyric acid synthesis   \n",
       "3      Gamma-aminobutyric acid synthesis   \n",
       "4      Gamma-aminobutyric acid synthesis   \n",
       "...                                  ...   \n",
       "32985                                NaN   \n",
       "32986                                NaN   \n",
       "32987                                NaN   \n",
       "32988                                NaN   \n",
       "32989                                NaN   \n",
       "\n",
       "                                            MEDIATOR_IDS  \n",
       "0      http://pathwaycommons.org/pc13/Catalysis_c5d26...  \n",
       "1      http://pathwaycommons.org/pc13/Catalysis_80d0d...  \n",
       "2      http://pathwaycommons.org/pc13/Catalysis_c5d26...  \n",
       "3      http://pathwaycommons.org/pc13/Catalysis_c5d26...  \n",
       "4      http://pathwaycommons.org/pc13/Catalysis_c5d26...  \n",
       "...                                                  ...  \n",
       "32985                                                NaN  \n",
       "32986                                                NaN  \n",
       "32987                                                NaN  \n",
       "32988                                                NaN  \n",
       "32989                                                NaN  \n",
       "\n",
       "[32990 rows x 7 columns]"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('../C5_PathwayCommons13.panther.hgnc.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a portion of Pathway Commons containing interactions from known pathways (Kyoto Encylopedia of Genes and Genomes65, NetPath66, PANTHER67 and WikiPathways68) (9,344 p, 242,962 i), #!need to find out\n",
    "\n",
    "with open('../m2.cp.wikipathways.v2022.1.Mm.entrez.gmt') as gmt:\n",
    "    print(gmt.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a network of metabolic genes based on Recon (v.2, http://vmh.uni.lu) (1,628 p, 246,937 i). #! recon https://www.vmh.life/#home matlab file?\n",
    "import scipy.io\n",
    "Recon3DModel_mat =  ('../Recon3D_301/Recon3DModel_301.mat')\n",
    "Recon3D_mat = scipy.io.loadmat('../Recon3D_301/Recon3D_301.mat')\n",
    "\n",
    "# To build this last network, we linked two metabolic proteins (enzymes or transporters) when the product metabolite of the first was the substrate of the second, \n",
    "# or when both were needed to perform a certain reaction, suggesting that they are part of the same protein complex. \n",
    "# Edges between proteins were weighted inversely proportional to the number of reactions involving their shared metabolites, \n",
    "# so that ‘currency’ metabolites such as ATP and water had marginal impact on the network connectivity. To control for indirect associations, \n",
    "# we deconvoluted the network using edge weights and setting a network deconvolution score cutoff of 0.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['__header__', '__version__', '__globals__', 'Recon3D'])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Recon3D_mat.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('DB/recon_3d.txt', 'w') as f:\n",
    "    writer = csv.writer(f)\n",
    "    for x in range(1,len(Recon3D_mat['Recon3D'][0][0])):\n",
    "        writer.writerow(Recon3D_mat['Recon3D'][0][0][x].tolist())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Recon3D_mat['Recon3D'][0][0][6].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[array(['10FTHF5GLUtl'], dtype='<U12')],\n",
       "       [array(['10FTHF5GLUtm'], dtype='<U12')],\n",
       "       [array(['10FTHF6GLUtl'], dtype='<U12')],\n",
       "       ...,\n",
       "       [array(['NADH2_u10mi'], dtype='<U11')],\n",
       "       [array(['CYOOm3i'], dtype='<U7')],\n",
       "       [array(['CYOOm2i'], dtype='<U7')]], dtype=object)"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Recon3D_mat['Recon3D'][0][0][4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8399"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Recon3D_mat['Recon3D'][0][0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Recon3DModel_mat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "#network processing\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hot net"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://pypi.org/project/decorator/3.3.2/ \n",
    "# install decorator"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.6 ('ml-air')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "894005d3a7f9e7f90f1d225b789862a942f5be96f0c7d0676c3b66e7cb3c009e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
